// @voicify/react-native-voicify-assistant-bare/src/models/customAssistantApiModels.ts
export interface CustomAssistantRequest {
    requestId: string
    context: CustomAssistantRequestContext
    device: CustomAssistantDevice
    user: CustomAssistantUser
}

export type CustomAssistantRequestType = "IntentRequest" | string 
export interface CustomAssistantRequestContext {
    sessionId: string
    noTracking?: boolean
    requestType: CustomAssistantRequestType
    requestName?: string
    slots?: {[key: string]: string;}
    originalInput: string
    channel: string
    requiresLanguageUnderstanding: boolean
    locale: string
    additionalRequestAttributes?: {[key: string]: any;}
    additionalSessionAttributes?: {[key: string]: any;}
    additionalSessionFlags?: string[]
}

export interface CustomAssistantDevice {
    id: string
    name: string
    supportsVideo?: boolean
    supportsForegroundImage?: boolean
    supportsBackgroundImage?: boolean
    supportsAudio?: boolean
    supportsSsml?: boolean
    supportsDisplayText?: boolean
    supportsVoiceInput?: boolean
    supportsTextInput?: boolean
}

export interface CustomAssistantUser {
    id: string
    name?: string
    accessToken?: string
    additionalUserAttributes?: { [key: string]: any; }
    additionalUserFlags?: string[]
}


export interface CustomAssistantResponse {
    responseId: string
    ssml?: string
    outputSpeech?: string
    displayText?: string
    displayTitle?: string
    responseTemplate?: string
    foregroundImage?: MediaItemModel
    backgroundImage?: MediaItemModel
    audioFile?: MediaItemModel
    videoFile?: MediaItemModel
    sessionAttributes?: {[key: string]: any;}
    effects?: VoicifySessionEffect[]
    hints?: string[]
    listItems?: CustomAssistantListItem[]
    endSession: boolean
}

export interface CustomAssistantListItem {
    id?: string
    title?: string
    description?: string
    image?: MediaItemModel
}

export interface MediaItemModel {
    id?: string
    url?: string
}

export interface VoicifySessionData {
    id?: string
    sessionFlags?: string[],
    sessionAttributes?: {[key: string]: any;}
}

export interface VoicifyUserData {
    id?: string
    userFlags?: string[],
    userAttributes?: {[key: string]: any;}
}

export interface VoicifySessionEffect {
    id?: string;
    name?: string;
    requestId?: string;
    effectName?: string;
    data: any
}

export interface TTSRequest {
    applicationId: string 
    applicationSecret: string
    ssmlRequest: SsmlRequest
}

export interface SsmlRequest {
    ssml: string
    locale: string
    voice: string
}

export interface TTSData {
    rootElementType: string,
    url: string
}
